### PR TITLE
Fixed 47 missing dependencies and 51 excessive dependencies in Makefile

### PR DESCRIPTION
**Pull request by vemakereporter**
Hi, I've fixed 47 missing dependencies and 51 excessive dependencies reported.
Those issues can cause incorrect results or unnecessary rebuilds when capstone is incrementally built.

For example, any changes in "arch/AArch64/AArch64BaseInfo.h" will not cause "arch/AArch64/AArch64BaseInfo.o" to be rebuilt, which is incorrect.

"MCInstrDesc.h" is specified as the prerequisite of "arch/Mips/MipsDisassembler.o". However, this header file is not read by the process when "arch/Mips/MipsDisassembler.o" is built. Hence, the changes in "MCInstrDesc.h" can cause an unnecessary rebuild of "arch/Mips/MipsDisassembler.o", which slow down the build of the whole project.  

A useful tool is the -M class of flags that we can pass to clang/gcc to generate "dependency files". Even if we don't use the flag directly in the Makefile, a script that we use could take advantage of the feature.
I modify the Makefile to read dependencies from compiler-generated files. I've checked. The dependences generate from the new Makefile are correct.
Looking forward to your confirmation.

Thanks
Vemake

### CODE REVIEW

**Comment by vemakereporter** [comment]
@aquynh @tmfink I have resubmitted the new PR to `next` track. Waiting for your confirmation.

**Comment by tmfink** [comment]
@vemakereporter After doing a make and make clean, there are still `*.d` files in `tests/` and `suite/fuzz/`.

~~~
$ make && make clean
$ find . -type f -name '*.d'
./tests/test_mos65xx.d
./tests/test_m68k.d
./tests/test_iter.d
./tests/test_systemz.d
./tests/test_arm.d
./tests/test_mips.d
./tests/test_evm.d
./tests/test_wasm.d
./tests/test_xcore.d
./tests/test_x86.d
./tests/test_skipdata.d
./tests/test_riscv.d
./tests/test_ppc.d
./tests/test_sparc.d
./tests/test_detail.d
./tests/test_basic.d
./tests/test_m680x.d
./tests/test_bpf.d
./tests/test_tms320c64x.d
./tests/test_arm64.d
./tests/test_customized_mnem.d
./suite/fuzz/drivermc.d
./suite/fuzz/fuzz_decode_platform.d
./suite/fuzz/fuzz_disasm.d
./suite/fuzz/driverbin.d
./suite/fuzz/platform.d
~~~

**Comment by tmfink** [comment]
For tracking purposes, this is a follow-up to PR #1521 to the `next` branch.

File: Makefile [hunk 1]
```diff
@@ -444,6 +444,16 @@ else
 	$(generate-pkgcfg)
 endif
 
+# destination path macro we'll use below
+df = $(*F)
```
Comment (tmfink) [inline Makefile]: I don't see this used

**Comment by vemakereporter** [comment]
> @vemakereporter After doing a make and make clean, there are still `*.d` files in `tests/` and `suite/fuzz/`.
> 
> ```
> $ make && make clean
> $ find . -type f -name '*.d'
> ./tests/test_mos65xx.d
> ./tests/test_m68k.d
> ./tests/test_iter.d
> ./tests/test_systemz.d
> ./tests/test_arm.d
> ./tests/test_mips.d
> ./tests/test_evm.d
> ./tests/test_wasm.d
> ./tests/test_xcore.d
> ./tests/test_x86.d
> ./tests/test_skipdata.d
> ./tests/test_riscv.d
> ./tests/test_ppc.d
> ./tests/test_sparc.d
> ./tests/test_detail.d
> ./tests/test_basic.d
> ./tests/test_m680x.d
> ./tests/test_bpf.d
> ./tests/test_tms320c64x.d
> ./tests/test_arm64.d
> ./tests/test_customized_mnem.d
> ./suite/fuzz/drivermc.d
> ./suite/fuzz/fuzz_decode_platform.d
> ./suite/fuzz/fuzz_disasm.d
> ./suite/fuzz/driverbin.d
> ./suite/fuzz/platform.d
> ```

test and suite/fuzz have their separate Makefiles. I include a removal command in their clean targets too.

**Comment by vemakereporter** [comment]
@tmfink @aquynh

File: suite/fuzz/Makefile [hunk 2]
```diff
@@ -53,6 +53,7 @@ all: $(REPRODUCERMC) $(REPRODUCERBIN) $(FUZZERBIN) $(PLATFORMDECODE)
 
 clean:
 	rm -rf fuzz_harness $(OBJS) $(PLATFORMDECODE) $(REPRODUCERMC) $(REPRODUCERBIN) $(FUZZERBIN) $(OBJDIR)/lib$(LIBNAME).* $(OBJDIR)/$(LIBNAME).*
+	rm -rf *.d $(OBJDIR)/*.d
```
Comment (tmfink) [inline suite/fuzz/Makefile]: I don't think you want the '-r' day here

File: tests/Makefile [hunk 3]
```diff
@@ -137,6 +137,7 @@ all: $(BINARY)
 
 clean:
 	rm -rf $(OBJS) $(BINARY) $(TESTDIR)/*.exe $(TESTDIR)/*.static $(OBJDIR)/lib$(LIBNAME).* $(OBJDIR)/$(LIBNAME).*
+	rm -rf *.d $(TESTDIR)/*.d $(OBJDIR)/*.d
```
Comment (tmfink) [inline tests/Makefile]: Same as above

File: suite/fuzz/Makefile (hunk 2 above)
Comment (vemakereporter) [reply suite/fuzz/Makefile]: Yes, you are right

File: tests/Makefile (hunk 3 above)
Comment (vemakereporter) [reply tests/Makefile]: Yes, you are right.

**Comment by vemakereporter** [comment]
@tmfink @aquynh Is it okay?

**Comment by vemakereporter** [comment]
@tmfink @aquynh We have tested the project again, and found that these header files cause 51 excessive dependencies in total, which cause unnecessary incremental rebuilds of 51 targets including "arch/Mips/MipsDisassembler.o". Looking forward to your confirmation.

File: cstool/Makefile [hunk 4]
```diff
@@ -37,6 +37,7 @@ endif
 
 clean:
 	${RM} -rf *.o $(TARGET)
+	${RM} -rf *.d
```
Comment (aquynh) [inline cstool/Makefile]: there is no more .o files?

File: Makefile [hunk 5]
```diff
@@ -444,6 +444,13 @@ else
 	$(generate-pkgcfg)
 endif
 
+# create a list of auto dependencies
+AUTODEPS:= $(patsubst %.o,%.d, $(LIBOBJ))
+
```
Comment (aquynh) [inline Makefile]: please remove 1 blank line here

File: cstool/Makefile (hunk 4 above)
Comment (vemakereporter) [reply cstool/Makefile]: Yes, it just creates .d files aside to the .o files.

File: Makefile (hunk 5 above)
Comment (vemakereporter) [reply Makefile]: Done

**Comment by vemakereporter** [comment]
@tmfink @aquynh We have fixed it.

**Comment by aquynh** [comment]
cool, just 1 minor issue now: we have a conflict in Makefile, so i cannot merge. could you please fix it?

**Comment by vemakereporter** [comment]
> cool, just 1 minor issue now: we have a conflict in Makefile, so i cannot merge. could you please fix it?

Conflict resolved.

**Comment by aquynh** [comment]
merged, thanks!

**Pull request merged**